### ISSUE TITLE
sql: internal error: top-level relational expression cannot have outer columns

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#41820]
```
CREATE TABLE IF NOT EXISTS seed AS
        SELECT
                g AS _int8, g AS _float4, g AS _float8
        FROM
                generate_series(NULL, NULL) AS g;

SELECT
        NULL
FROM
        seed AS tab_27687
HAVING
        0
        IN (
                        SELECT
                                tab_27693._int8
                        FROM
                                seed AS tab_27691
                                JOIN (VALUES (tab_27687._float8))
                                                AS tab_27692 (col_53475) ON
                                                tab_27691._float4
                                                = tab_27692.col_53475
                                                AND tab_27691._float8
                                                        = tab_27692.col_53475
                                JOIN seed AS tab_27693 ON true
                );
```

```
pq: internal error: top-level relational expression cannot have outer columns: (3)
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:218: Optimize()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:424: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:162: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:743: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:625: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1244: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1173: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:444: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584: func1()
```

**Comment by rytaft** [comment]
Sorry, I'm not sure why I never got an email with your ping. It seems that the Github email integration must have been broken somehow. Anyway, adding this to my to-do list now.

**Comment by rytaft** [comment]
It seems that this particular case was fixed by #41033.

But there is another example from the community slack that still fails:
```
create table c1 (id int primary key, text_column string);
create table c2 (id int primary key, text_column string);
create table c3 (id int primary key, text_column string);
select json_agg((select (row_to_json(( select c2.id, min(c2.text_column), json_agg(( select row_to_json(( select c3.id, c3.text_column)) )) )) ))) from c1, c2, c3 group by c1.id, c2.id, c3.id;
```

**Comment by rytaft** [comment]
An even simpler version that causes a slightly different error:
```
SELECT count((SELECT min(c2.text_column)))
FROM c2
GROUP BY c2.id;
```
I think the problem with all of these is that the optimizer should reject them for containing nested aggregates (that's what Postgres does, anyway), but it can't recognize that the aggregates are nested in this case.
